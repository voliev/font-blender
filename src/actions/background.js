import { SET_BACKGROUND_COLOR } from "../constants";

export const setBackgroundColor = (color = "#fff") => ({
  type: SET_BACKGROUND_COLOR,
  color
});
