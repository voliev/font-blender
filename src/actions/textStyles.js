import { SET_STYLES_INHERITANCE, UPDATE_ELEMENT_STYLES } from "../constants";

export const setStylesInheritance = element => ({
  type: SET_STYLES_INHERITANCE,
  element
});

export const updateElementStyles = (element, property, value) => ({
  type: UPDATE_ELEMENT_STYLES,
  element,
  property,
  value
});
