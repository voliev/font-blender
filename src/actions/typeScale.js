import {
  SET_DEFAULT_TYPE_SCALE_STATE,
  SET_FONT_SIZE,
  SET_TYPE_SCALE
} from "../constants";

export const setTypeScale = (scale = 1.414) => ({
  type: SET_TYPE_SCALE,
  scale
});

export const setFontSize = (fontSize = 16) => ({
  type: SET_FONT_SIZE,
  fontSize
});

export const setDefaultTypeScaleState = () => ({
  type: SET_DEFAULT_TYPE_SCALE_STATE
});
