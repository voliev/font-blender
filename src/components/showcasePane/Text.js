import React from "react";
import PropTypes from "prop-types";
import { isPropTypeHexColor } from "../../utils";
import { ELEMENTS } from "../../constants";
import "./styles/Text.css";

const Text = ({ background, elementsStyles }) => (
  <div
    className="Text"
    style={{
      backgroundColor: `${background.color}`,
      ...elementsStyles.body
    }}
  >
    <div className="text-wrapper">
      <h1 style={elementsStyles.h1}>Lorem ipsum dolor sit amet</h1>

      <p style={elementsStyles.p}>
        Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo
        ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis
        dis parturient montes, nascetur ridiculus mus. Donec quam felis,
        ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa
        quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget,
        arcu.
      </p>

      <h2 style={elementsStyles.h2}>Sapiente sed voluptatibus neque</h2>

      <p style={elementsStyles.p}>
        In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo. Nullam
        dictum felis eu pede mollis pretium. Integer tincidunt. Cras dapibus.
        Vivamus elementum semper nisi. Aenean vulputate eleifend tellus. Aenean
        leo ligula, porttitor eu, consequat vitae, eleifend ac, enim. Aliquam
        lorem ante, dapibus in, viverra quis, feugiat a, tellus. Phasellus
        viverra nulla ut metus varius laoreet. Quisque rutrum.
      </p>

      <ul style={elementsStyles.li}>
        <li>Lorem ipsum</li>
        <li>Dolor sit</li>
        <li>Adipisicing elit</li>
        <li>Doloribus</li>
        <li>Odio praesentium</li>
        <li>Consectetur</li>
      </ul>

      <p style={elementsStyles.p}>
        Aenean imperdiet. Etiam ultricies nisi vel augue. Curabitur ullamcorper
        ultricies nisi. Nam eget dui. Etiam rhoncus. Maecenas tempus, tellus
        eget condimentum rhoncus, sem quam semper libero, sit amet adipiscing
        sem neque sed ipsum. Nam quam nunc, blandit vel, luctus pulvinar,
        hendrerit id, lorem. Maecenas nec odio et ante tincidunt tempus. Donec
        vitae sapien ut libero venenatis faucibus.
      </p>

      <h3 style={elementsStyles.h3}>Corrupti quia optio maiores*</h3>

      <p style={elementsStyles.p}>
        Cras ultricies mi eu turpis hendrerit fringilla. Vestibulum ante ipsum
        primis in faucibus orci luctus et ultrices posuere cubilia Curae; In ac
        dui quis mi consectetuer lacinia. Nam pretium turpis et arcu. Duis arcu
        tortor, suscipit eget, imperdiet nec, imperdiet iaculis, ipsum. Sed
        aliquam ultrices mauris. Integer ante arcu, accumsan a, consectetuer
        eget, posuere ut, mauris. Praesent adipiscing.
      </p>

      <h4 style={elementsStyles.h4}>Dolorum labore accusamus dolor</h4>

      <p style={elementsStyles.p}>
        Nullam quis ante. Etiam sit amet orci eget eros faucibus tincidunt. Duis
        leo. Sed fringilla mauris sit amet nibh. Donec sodales sagittis magna.
        Sed consequat, leo eget bibendum sodales, augue velit cursus nunc, quis
        gravida magna mi a libero. Fusce vulputate eleifend sapien. Vestibulum
        purus quam, scelerisque ut, mollis sed, nonummy id, metus. Nullam
        accumsan lorem in dui.
      </p>

      <h4 style={elementsStyles.h4}>Fuga autem rerum</h4>

      <p style={elementsStyles.p}>
        Phasellus ullamcorper ipsum rutrum nunc. Nunc nonummy metus. Vestibulum
        volutpat pretium libero. Cras id dui. Aenean ut eros et nisl sagittis
        vestibulum. Nullam nulla eros, ultricies sit amet, nonummy id, imperdiet
        feugiat, pede. Sed lectus. Donec mollis hendrerit risus. Phasellus nec
        sem in justo pellentesque facilisis. Etiam imperdiet imperdiet orci.
        Nunc nec neque.
      </p>

      <h4 style={elementsStyles.h4}>Sunt facere reprehenderit facilis</h4>

      <p style={elementsStyles.p}>
        Phasellus leo dolor, tempus non, auctor et, hendrerit quis, nisi.
        Curabitur ligula sapien, tincidunt non, euismod vitae, posuere
        imperdiet, leo. Maecenas malesuada. Praesent congue erat at massa. Sed
        cursus turpis vitae tortor. Donec posuere vulputate arcu. Phasellus
        accumsan cursus velit. Vestibulum ante ipsum primis in faucibus orci
        luctus et ultrices posuere cubilia Curae; Sed aliquam, nisi quis
        porttitor congue, elit erat euismod orci, ac placerat dolor lectus quis
        orci.
      </p>

      {
        <p>
          <small style={elementsStyles.small}>
            * Fusce id purus. Ut varius tincidunt libero. Phasellus dolor.
            Maecenas vestibulum mollis
          </small>
        </p>
      }
    </div>
  </div>
);

Text.propTypes = {
  elementsStyles: PropTypes.exact(
    ELEMENTS.reduce((checks, element) => {
      checks[element] = PropTypes.exact({
        color: isPropTypeHexColor,
        fontFamily: PropTypes.string.isRequired,
        fontSize: PropTypes.string.isRequired,
        fontStyle: PropTypes.string.isRequired,
        fontWeight: PropTypes.number.isRequired
      });

      return checks;
    }, {})
  ),
  background: PropTypes.exact({
    color: isPropTypeHexColor
  }).isRequired
};

export default Text;
