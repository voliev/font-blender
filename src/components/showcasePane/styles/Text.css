.Text .text-wrapper {
  margin: 0 auto;
  padding: var(--space-xl) var(--space-md) var(--space-xxl) var(--space-md);
  width: 50rem;
}
