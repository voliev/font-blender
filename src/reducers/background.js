import { SET_BACKGROUND_COLOR, SET_DEFAULT_STATE } from "../constants";

const initialState = {
  color: "#fff"
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_BACKGROUND_COLOR:
      return {
        ...state,
        color: action.color
      };
    case SET_DEFAULT_STATE:
      return initialState;
    default:
      return state;
  }
};

export const getBackground = state => state.styles.present.background;
export const getBackgroundDefaultState = () => ({ ...initialState });
