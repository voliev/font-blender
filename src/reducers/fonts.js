import { combineReducers } from "redux";
import {
  FETCH_FONTS_DATA,
  FETCH_FONTS_DATA_FAILURE,
  FETCH_FONTS_DATA_SUCCESS,
  SET_FONTS,
  UPDATE_FONT
} from "../constants";

export const fontsDataStatus = (state = "", action) => {
  switch (action.type) {
    case FETCH_FONTS_DATA:
      return "fetching";
    case FETCH_FONTS_DATA_SUCCESS:
      return "success";
    case FETCH_FONTS_DATA_FAILURE:
      return "failure";
    default:
      return state;
  }
};

const initialState = {};

export const all = (state = initialState, action) => {
  switch (action.type) {
    case SET_FONTS:
      return action.fonts;
    case UPDATE_FONT:
      return {
        ...state,
        [action.family]: {
          ...state[action.family],
          ...action.updates
        }
      };
    default:
      return state;
  }
};

export default combineReducers({
  all,
  fontsDataStatus
});

export const getFonts = state => state.fonts.all;
export const getFont = (state, family) => state.fonts.all[family];
export const getFontsDataStatus = state => state.fonts.fontsDataStatus;
