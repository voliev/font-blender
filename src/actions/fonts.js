import WebFont from "webfontloader";
import { getFonts } from "../reducers/fonts";
import {
  GOOGLE_FONTS_API_ENDPOINT,
  FETCH_FONTS_DATA,
  FETCH_FONTS_DATA_FAILURE,
  FETCH_FONTS_DATA_SUCCESS,
  SET_FONTS,
  UPDATE_FONT
} from "../constants";

export const fetchFontsData = () => ({
  type: FETCH_FONTS_DATA
});

export const fetchFontsDataSuccess = () => ({
  type: FETCH_FONTS_DATA_SUCCESS
});

export const fetchFontsDataFailure = () => ({
  type: FETCH_FONTS_DATA_FAILURE
});

export const setFonts = fonts => ({
  type: SET_FONTS,
  fonts
});

export const startFetchFontsData = () => {
  return async dispatch => {
    const apiRequestURL = `${GOOGLE_FONTS_API_ENDPOINT}?key=${
      process.env.REACT_APP_GOOGLE_FONTS_API_KEY
    }&sort=popularity`;

    dispatch(fetchFontsData());

    try {
      const res = await fetch(apiRequestURL);
      const { items } = await res.json();
      dispatch(fetchFontsDataSuccess());

      const fonts = items.reduce((result, font) => {
        result[font.family] = {
          family: font.family,
          category: font.category,
          variants: font.variants,
          subsets: font.subsets,
          fetching: false,
          active: false,
          requested: false,
          prefetched: false,
          fetched: false
        };

        return result;
      }, {});

      return dispatch(setFonts(fonts));
    } catch (e) {
      dispatch(fetchFontsDataFailure());
      // TODO: Eject and replace service worker with custom one.
      //       Implement fonts' json data caching.
      return dispatch(setFonts({}));
    }
  };
};

export const updateFont = (family, updates) => ({
  type: UPDATE_FONT,
  family,
  updates
});

export const startFetchFont = (family, charRange = undefined) => {
  return (dispatch, getState) => {
    const state = getState();
    const fonts = getFonts(state);
    const font = fonts[family];
    const fetchFull = !charRange;
    const requestString = generateRequestString(font, fetchFull);
    const options = generateRequestOptions(
      requestString,
      charRange,
      dispatch,
      state
    );

    WebFont.load(options);

    const updatedState = getState();
    const updatedFonts = getFonts(updatedState);

    // For "then" chaining in selectedFonts.js async action creator
    return Promise.resolve(updatedFonts[family]);
  };
};

export const generateRequestString = (font, fetchFull = false) => {
  const { family, subsets, variants } = font;

  if (fetchFull) {
    return `${family}:${variants.join(",")}`;
  }

  const weight = variants.includes("regular")
    ? "400"
    : parseInt(variants[0], 10).toString();
  const subset = subsets.includes("latin") ? "latin" : subsets[0];

  return `${family}:${weight}:${subset}`;
};

export const generateRequestOptions = (request, chars, dispatch) => ({
  google: {
    families: [request],
    text: chars
  },
  classes: false,
  timeout: 10000,
  fontloading: getEventListener("fontloading", dispatch, chars),
  fontactive: getEventListener("fontactive", dispatch, chars),
  fontinactive: getEventListener("fontinactive", dispatch, chars)
});

export const getEventListener = (event, dispatch, charRange) => {
  const listeners = {
    fontloading: family => {
      const updates = Object.assign(
        { fetching: true },
        charRange ? { requested: true } : {}
      );
      dispatch(updateFont(family, updates));
    },
    fontactive: family => {
      const updates = Object.assign(
        { active: true, fetching: false },
        charRange ? { prefetched: true } : { fetched: true }
      );
      dispatch(updateFont(family, updates));
    },
    fontinactive: family => {
      const updates = Object.assign(
        { fetching: false },
        charRange ? { active: false } : {}
      );
      dispatch(updateFont(family, updates));
    }
  };

  return listeners[event];
};
