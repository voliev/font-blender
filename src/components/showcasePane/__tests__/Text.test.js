import React from "react";
import { mount, shallow } from "enzyme";
import Text from "../Text";
import { generateElementsStyles } from "../../../utils";
import fonts from "../../../fixtures/fonts";
import background from "../../../fixtures/background";
import typeScale from "../../../fixtures/typeScale";
import textStyles from "../../../fixtures/textStyles";

const setup = (renderFn = shallow, propsOverrides = {}) => {
  const props = {
    background,
    elementsStyles: generateElementsStyles(typeScale, textStyles, fonts),
    ...propsOverrides
  };

  return {
    ...props,
    wrapper: renderFn(<Text {...props} />)
  };
};

it("should render without crashing", () => {
  const { wrapper } = setup();

  expect(wrapper).toMatchSnapshot();
});

it("should set corresponding styles", () => {
  const { elementsStyles, wrapper } = setup(mount);

  expect(wrapper.find(".Text").prop("style")).toEqual({
    backgroundColor: `${background.color}`,
    ...elementsStyles.body
  });

  expect(wrapper.find("h1").prop("style")).toEqual(elementsStyles.h1);
  expect(wrapper.find("h2").prop("style")).toEqual(elementsStyles.h2);
  expect(wrapper.find("h3").prop("style")).toEqual(elementsStyles.h3);
  wrapper.find("h4").forEach(node => {
    expect(node.prop("style")).toEqual(elementsStyles.h4);
  });
  expect(wrapper.find("ul").prop("style")).toEqual(elementsStyles.li);
  expect(wrapper.find("small").prop("style")).toEqual(elementsStyles.small);
});
