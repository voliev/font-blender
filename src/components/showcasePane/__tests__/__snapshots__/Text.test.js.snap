// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without crashing 1`] = `
<div
  className="Text"
  style={
    Object {
      "backgroundColor": "#fff",
      "color": "#333",
      "fontFamily": "sans-serif",
      "fontSize": "16.00px",
      "fontStyle": "normal",
      "fontWeight": 400,
    }
  }
>
  <div
    className="text-wrapper"
  >
    <h1
      style={
        Object {
          "color": "#333",
          "fontFamily": "sans-serif",
          "fontSize": "63.96px",
          "fontStyle": "normal",
          "fontWeight": 400,
        }
      }
    >
      Lorem ipsum dolor sit amet
    </h1>
    <p>
      Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget, arcu.
    </p>
    <h2
      style={
        Object {
          "color": "#333",
          "fontFamily": "sans-serif",
          "fontSize": "45.23px",
          "fontStyle": "normal",
          "fontWeight": 400,
        }
      }
    >
      Sapiente sed voluptatibus neque
    </h2>
    <p>
      In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo. Nullam dictum felis eu pede mollis pretium. Integer tincidunt. Cras dapibus. Vivamus elementum semper nisi. Aenean vulputate eleifend tellus. Aenean leo ligula, porttitor eu, consequat vitae, eleifend ac, enim. Aliquam lorem ante, dapibus in, viverra quis, feugiat a, tellus. Phasellus viverra nulla ut metus varius laoreet. Quisque rutrum.
    </p>
    <ul
      style={
        Object {
          "color": "#333",
          "fontFamily": "sans-serif",
          "fontSize": "16.00px",
          "fontStyle": "normal",
          "fontWeight": 400,
        }
      }
    >
      <li>
        Lorem ipsum
      </li>
      <li>
        Dolor sit
      </li>
      <li>
        Adipisicing elit
      </li>
      <li>
        Doloribus
      </li>
      <li>
        Odio praesentium
      </li>
      <li>
        Consectetur
      </li>
    </ul>
    <p>
      Aenean imperdiet. Etiam ultricies nisi vel augue. Curabitur ullamcorper ultricies nisi. Nam eget dui. Etiam rhoncus. Maecenas tempus, tellus eget condimentum rhoncus, sem quam semper libero, sit amet adipiscing sem neque sed ipsum. Nam quam nunc, blandit vel, luctus pulvinar, hendrerit id, lorem. Maecenas nec odio et ante tincidunt tempus. Donec vitae sapien ut libero venenatis faucibus.
    </p>
    <h3
      style={
        Object {
          "color": "#333",
          "fontFamily": "sans-serif",
          "fontSize": "31.99px",
          "fontStyle": "normal",
          "fontWeight": 400,
        }
      }
    >
      Corrupti quia optio maiores*
    </h3>
    <p>
      Cras ultricies mi eu turpis hendrerit fringilla. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; In ac dui quis mi consectetuer lacinia. Nam pretium turpis et arcu. Duis arcu tortor, suscipit eget, imperdiet nec, imperdiet iaculis, ipsum. Sed aliquam ultrices mauris. Integer ante arcu, accumsan a, consectetuer eget, posuere ut, mauris. Praesent adipiscing.
    </p>
    <h4
      style={
        Object {
          "color": "#333",
          "fontFamily": "sans-serif",
          "fontSize": "22.62px",
          "fontStyle": "normal",
          "fontWeight": 400,
        }
      }
    >
      Dolorum labore accusamus dolor
    </h4>
    <p>
      Nullam quis ante. Etiam sit amet orci eget eros faucibus tincidunt. Duis leo. Sed fringilla mauris sit amet nibh. Donec sodales sagittis magna. Sed consequat, leo eget bibendum sodales, augue velit cursus nunc, quis gravida magna mi a libero. Fusce vulputate eleifend sapien. Vestibulum purus quam, scelerisque ut, mollis sed, nonummy id, metus. Nullam accumsan lorem in dui.
    </p>
    <h4
      style={
        Object {
          "color": "#333",
          "fontFamily": "sans-serif",
          "fontSize": "22.62px",
          "fontStyle": "normal",
          "fontWeight": 400,
        }
      }
    >
      Fuga autem rerum
    </h4>
    <p>
      Phasellus ullamcorper ipsum rutrum nunc. Nunc nonummy metus. Vestibulum volutpat pretium libero. Cras id dui. Aenean ut eros et nisl sagittis vestibulum. Nullam nulla eros, ultricies sit amet, nonummy id, imperdiet feugiat, pede. Sed lectus. Donec mollis hendrerit risus. Phasellus nec sem in justo pellentesque facilisis. Etiam imperdiet imperdiet orci. Nunc nec neque.
    </p>
    <h4
      style={
        Object {
          "color": "#333",
          "fontFamily": "sans-serif",
          "fontSize": "22.62px",
          "fontStyle": "normal",
          "fontWeight": 400,
        }
      }
    >
      Sunt facere reprehenderit facilis
    </h4>
    <p>
      Phasellus leo dolor, tempus non, auctor et, hendrerit quis, nisi. Curabitur ligula sapien, tincidunt non, euismod vitae, posuere imperdiet, leo. Maecenas malesuada. Praesent congue erat at massa. Sed cursus turpis vitae tortor. Donec posuere vulputate arcu. Phasellus accumsan cursus velit. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Sed aliquam, nisi quis porttitor congue, elit erat euismod orci, ac placerat dolor lectus quis orci.
    </p>
    <p>
      <small
        style={
          Object {
            "color": "#333",
            "fontFamily": "sans-serif",
            "fontSize": "11.32px",
            "fontStyle": "normal",
            "fontWeight": 400,
          }
        }
      >
        * Fusce id purus. Ut varius tincidunt libero. Phasellus dolor. Maecenas vestibulum mollis
      </small>
    </p>
  </div>
</div>
`;
